File: runtime/go/expr.go
package uiutils

import "strings"

// Tiny attribute-logic helpers used by generated components. One
// implementation per runtime family keeps every default-resolution decision
// table identical across Templ, React, Svelte, and Vue (mirrors expr.ts).

// TrimStr trims surrounding whitespace.
func TrimStr(v string) string { return strings.TrimSpace(v) }

// IsSetStr reports whether the trimmed string is non-empty.
func IsSetStr(v string) bool { return strings.TrimSpace(v) != "" }

// DefaultIfEmpty returns fallback when the trimmed value is empty.
func DefaultIfEmpty(v, fallback string) string {
	t := strings.TrimSpace(v)
	if t == "" {
		return fallback
	}
	return t
}

// OneOfOr returns the trimmed value when in allowed, otherwise fallback.
func OneOfOr(v, fallback string, allowed ...string) string {
	t := strings.TrimSpace(v)
	for _, a := range allowed {
		if t == a {
			return t
		}
	}
	return fallback
}

// OneOfOrLower lowercases before the whitelist check.
func OneOfOrLower(v, fallback string, allowed ...string) string {
	return OneOfOr(strings.ToLower(strings.TrimSpace(v)), fallback, allowed...)
}

// MapOr is a decision table: mapping[trim(v)] or fallback.
func MapOr(v string, mapping map[string]string, fallback string) string {
	if out, ok := mapping[strings.TrimSpace(v)]; ok {
		return out
	}
	return fallback
}

// MapOrLower lowercases before the table lookup.
func MapOrLower(v string, mapping map[string]string, fallback string) string {
	return MapOr(strings.ToLower(strings.TrimSpace(v)), mapping, fallback)
}

// IntMapOr is an integer decision table: mapping[v] or fallback.
func IntMapOr(v int, mapping map[int]string, fallback string) string {
	if out, ok := mapping[v]; ok {
		return out
	}
	return fallback
}

// BoolStr stringifies ARIA state: always "true"/"false", never omitted.
func BoolStr(v bool) string {
	if v {
		return "true"
	}
	return "false"
}

// ConcatTrim concatenates trimmed parts (icon Prefix + Name).
func ConcatTrim(parts ...string) string {
	var b strings.Builder
	for _, p := range parts {
		b.WriteString(strings.TrimSpace(p))
	}
	return b.String()
}

// If is an eager ternary for value selection in attribute expressions.
func If[T any](cond bool, then, els T) T {
	if cond {
		return then
	}
	return els
}

File: runtime/ts/expr.ts
/**
 * Tiny attribute-logic helpers used by generated components. One
 * implementation per runtime family keeps every default-resolution decision
 * table byte-identical across React, Svelte, and Vue (mirrors `expr.go`).
 */

export const trimStr = (v: string | number | null | undefined): string =>
  v === null || v === undefined ? "" : String(v).trim();

export const isSetStr = (v: string | null | undefined): boolean => trimStr(v) !== "";

export const isSetNum = (v: number | null | undefined): boolean => (v ?? 0) > 0;

/**
 * `trim(v) === "" ? fallback : trim(v)` — DefaultInputType-style defaults.
 * Generic so enum-typed props keep their literal union through the default.
 */
export const defaultIfEmpty = <T extends string>(v: T | null | undefined, fallback: T): T => {
  const t = trimStr(v) as T;
  return t === "" ? fallback : t;
};

/** Whitelist: returns `trim(v)` when in `allowed`, otherwise `fallback`. */
export const oneOfOr = <T extends string>(
  v: string | null | undefined,
  fallback: T,
  ...allowed: T[]
): T => {
  const t = trimStr(v) as T;
  return allowed.includes(t) ? t : fallback;
};

export const oneOfOrLower = <T extends string>(
  v: string | null | undefined,
  fallback: T,
  ...allowed: T[]
): T => {
  const t = trimStr(v).toLowerCase() as T;
  return allowed.includes(t) ? t : fallback;
};

/** Decision table: `mapping[trim(v)] ?? fallback`. */
export const mapOr = <T extends string>(
  v: string | null | undefined,
  mapping: Record<string, T>,
  fallback: T
): T => mapping[trimStr(v)] ?? fallback;

export const mapOrLower = <T extends string>(
  v: string | null | undefined,
  mapping: Record<string, T>,
  fallback: T
): T => mapping[trimStr(v).toLowerCase()] ?? fallback;

/** Integer decision table: `mapping[v] ?? fallback`. */
export const intMapOr = <T extends string>(
  v: number | null | undefined,
  mapping: Record<number, T>,
  fallback: T
): T => mapping[v ?? 0] ?? fallback;

/** Null-safe string equality (both sides normalized to ""). */
export const eqStr = (a: string | null | undefined, b: string | null | undefined): boolean =>
  (a ?? "") === (b ?? "");

/** ARIA state stringification: always "true"/"false", never omitted. */
export const boolStr = (v: boolean | null | undefined): "true" | "false" =>
  v ? "true" : "false";

/** Concatenation of trimmed parts (icon `Prefix + Name`). */
export const concatTrim = (...parts: (string | null | undefined)[]): string =>
  parts.map(trimStr).join("");

/** Omit-if-empty attribute value: `""` → `undefined` so the attr is dropped. */
export const orUndef = (v: string | null | undefined): string | undefined => {
  const t = trimStr(v);
  return t === "" ? undefined : String(v);
};

export type HeadingLevel = 1 | 2 | 3 | 4 | 5 | 6;

/** h1–h6 from a heading level; 0/undefined/out-of-range default to h2. */
export const titleTag = (order: HeadingLevel | number | null | undefined): string => {
  switch (order) {
    case 1:
      return "h1";
    case 3:
      return "h3";
    case 4:
      return "h4";
    case 5:
      return "h5";
    case 6:
      return "h6";
    default:
      return "h2";
  }
};
